### PR TITLE
Bug 2092041: Bump k8s to 1.24 and Golang to 1.18

### PR DESCRIPTION
**Pull request by candita**
Update to Golang 1.18
- .ci-operator.yaml
- Dockerfile
- go.mod

Update test/e2e/operator_test.go to use %v instead of %w in T.Errorf - needed under Go 1.18

Update these modules in go.mod to k8s 1.24.1:
- k8s.io/api v0.24.1
- k8s.io/apimachinery v0.24.1
- k8s.io/client-go v0.24.1
- k8s.io/kubectl v0.24.1

Update this module in go.mod to v0.12.1:
- sigs.k8s.io/controller-runtime v0.12.1

Everything else updated itself as a byproduct

### CODE REVIEW

**Review by frobware — COMMENTED** [review]
(I ran out of time today.) Just a comment regarding `logr` version change (v1.2.3 -> v1.2.0).

File: go.mod [hunk 1]
```diff
@@ -9,53 +9,62 @@ require (
...
+	github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578 // indirect
 	github.com/beorn7/perks v1.0.1 // indirect
-	github.com/cespare/xxhash/v2 v2.1.1 // indirect
+	github.com/cespare/xxhash/v2 v2.1.2 // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/emicklei/go-restful v2.9.5+incompatible // indirect
 	github.com/evanphx/json-patch v4.12.0+incompatible // indirect
 	github.com/fsnotify/fsnotify v1.5.1 // indirect
-	github.com/go-logr/logr v1.2.3 // indirect
+	github.com/go-logr/logr v1.2.0 // indirect
```
[the remark this answers is not in the document]
Comment (candita) [reply go.mod]: That was automatic, but I manually updated it to stay the same.  The code change was only comments.

[the remark this answers is not in the document]
Comment (frobware) [reply go.mod]: Manually editing the entry that go mod writes shouldn't be necessary. I followed your steps here: https://github.com/openshift/cluster-dns-operator/pull/329.

I pulled your PR and rebased against upstream/master. The diffs are:

```console
$ git diff --stat BZ-2092041 
 go.mod                                     |  2 +-
 go.sum                                     |  3 ++-
 vendor/golang.org/x/net/http2/errors.go    |  2 +-
 vendor/golang.org/x/net/http2/frame.go     |  3 ++-
 vendor/golang.org/x/net/http2/go118.go     | 17 +++++++++++++++++
 vendor/golang.org/x/net/http2/not_go118.go | 17 +++++++++++++++++
 vendor/golang.org/x/net/http2/transport.go | 51 +++++++++++++++++++++++++++++++++++++--------------
 vendor/modules.txt                         |  6 +++++-
 8 files changed, 82 insertions(+), 19 deletions(-)
```

# Differences

## go.mod

```
$ git diff BZ-2092041 go.mod
diff --git a/go.mod b/go.mod
index 60e3415d..949c56b8 100644
--- a/go.mod
+++ b/go.mod
@@ -49,7 +49,7 @@ require (
        github.com/prometheus/common v0.32.1 // indirect
        github.com/prometheus/procfs v0.7.3 // indirect
        github.com/spf13/pflag v1.0.5 // indirect
-       golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd // indirect
+       golang.org/x/net v0.0.0-20220403103023-749bd193bc2b // indirect
        golang.org/x/oauth2 v0.0.0-20211104180415-d3ed0bb246c8 // indirect
        golang.org/x/sys v0.0.0-20220209214540-3681064d5158 // indirect
        golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
```

## go.sum

```
$ git diff BZ-2092041 go.sum
diff --git a/go.sum b/go.sum
index e319fe06..15cd7d99 100644
--- a/go.sum
+++ b/go.sum
@@ -573,8 +573,9 @@ golang.org/x/net v0.0.0-20210805182204-aaa1db679c0d/go.mod h1:9nx3DQGgdP8bBQD5qx
 golang.org/x/net v0.0.0-20210825183410-e898025ed96a/go.mod h1:9nx3DQGgdP8bBQD5qxJ1jj9UTztislL4KSBs9R2vV5Y=
 golang.org/x/net v0.0.0-20211015210444-4f30a5c0130f/go.mod h1:9nx3DQGgdP8bBQD5qxJ1jj9UTztislL4KSBs9R2vV5Y=
 golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2/go.mod h1:9nx3DQGgdP8bBQD5qxJ1jj9UTztislL4KSBs9R2vV5Y=
-golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd h1:O7DYs+zxREGLKzKoMQrtrEacpb0ZVXA5rIwylE2Xchk=
 golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd/go.mod h1:CfG3xpIq0wQ8r1q4Su4UZFWDARRcnwPjda9FqA0JpMk=
+golang.org/x/net v0.0.0-20220403103023-749bd193bc2b h1:vI32FkLJNAWtGD4BwkThwEy6XS7ZLLMHkSkYfF8M0W0=
+golang.org/x/net v0.0.0-20220403103023-749bd193bc2b/go.mod h1:CfG3xpIq0wQ8r1q4Su4UZFWDARRcnwPjda9FqA0JpMk=
 golang.org/x/oauth2 v0.0.0-20180821212333-d2e6202438be/go.mod h1:N/0e6XlmueqKjAGxoOufVs8QHGRruUQn6yWY3a++T0U=
 golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421/go.mod h1:gOpvHmFTYa4IltrdGE7lF6nIHvwfUNPOp7c8zoXwtLw=
 golang.org/x/oauth2 v0.0.0-20190604053449-0f29369cfe45/go.mod h1:gOpvHmFTYa4IltrdGE7lF6nIHvwfUNPOp7c8zoXwtLw=
```

## vendor/modules.txt

```
$ git diff BZ-2092041 vendor/modules.txt
diff --git a/vendor/modules.txt b/vendor/modules.txt
index c433cc57..2e8269cb 100644
--- a/vendor/modules.txt
+++ b/vendor/modules.txt
@@ -74,6 +74,8 @@ github.com/google/gofuzz/bytesource
 # github.com/google/uuid v1.1.2
 ## explicit
 github.com/google/uuid
+# github.com/googleapis/gnostic v0.5.5
+## explicit; go 1.12
 # github.com/imdario/mergo v0.3.12
 ## explicit; go 1.13
 github.com/imdario/mergo
@@ -146,7 +148,7 @@ github.com/sirupsen/logrus
 # github.com/spf13/pflag v1.0.5
 ## explicit; go 1.12
 github.com/spf13/pflag
-# golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd
+# golang.org/x/net v0.0.0-20220403103023-749bd193bc2b
 ## explicit; go 1.17
 golang.org/x/net/context
 golang.org/x/net/context/ctxhttp
@@ -283,6 +285,8 @@ k8s.io/api/scheduling/v1beta1
 k8s.io/api/storage/v1
 k8s.io/api/storage/v1alpha1
 k8s.io/api/storage/v1beta1
+# k8s.io/apiextensions-apiserver v0.24.0
+## explicit; go 1.16
 # k8s.io/apimachinery v0.24.1
 ## explicit; go 1.16
 k8s.io/apimachinery/pkg/api/errors
```

I'll ignore the updates to `golang.org/x/net/http2`.

I'm surprised that you had to manually edit go.mod for the `logr` package.

**Comment by frobware** [comment]
/skip
/lgtm

Adding /hold for the observations in https://github.com/openshift/cluster-dns-operator/pull/328#discussion_r904843465.

/hold

**Pull request merged**